### PR TITLE
Update flake8 and mypy stuff

### PR DESCRIPTION
**Pull request by jakobandersen**
- Remove ``#noqa`` comments
- Switch comment-based type annotation to variable type hints
- Make mypy warn about redundant things
- General mypy fixes

There is still one error left:
```
breathe/renderer/sphinxrenderer.py:439: error: List item 0 has incompatible type "Tuple[str]"; expected "Tuple[str, str]"
```

### CODE REVIEW

File: breathe/directives/content_block.py [hunk 1]
```diff
@@ -103,10 +103,8 @@ class DoxygenNamespaceDirective(_DoxygenContentBlockDirective):
 
 class DoxygenGroupDirective(_DoxygenContentBlockDirective):
     kind = "group"
-    option_spec = {
-        **_DoxygenContentBlockDirective.option_spec,
-        "inner": flag,
-    }
+    option_spec = _DoxygenContentBlockDirective.option_spec.copy()
+    option_spec.update({"inner": flag})
```
Comment (michaeljones) [inline breathe/directives/content_block.py:107]: Out of interest, why do you prefer this formulation? Does it have an objective benefit or just preferred style?

**Review by michaeljones — APPROVED** [review]
Looks like a great improvement. Much respect for grinding through it.

**Comment by michaeljones** [comment]
@jakobandersen - please merge when you're happy with it :)

File: breathe/directives/content_block.py (hunk 1 above)
Comment (jakobandersen) [reply breathe/directives/content_block.py:107]: It is just a style choice, to be consistent with the equivalent cases I have seen in Sphinx to more quickly spot what is going on.

**Comment by jakobandersen** [comment]
Thanks, merged.

**Pull request merged**